src/pal/inc/pal.h

```diff
@@ -5894,206 +5894,83 @@ typedef EXCEPTION_DISPOSITION (*PFN_PAL_EXCEPTION_FILTER)(
     void *pvParam);
 
-struct _Unwind_Exception *PAL_TryExcept(
-    PFN_PAL_BODY pfnBody,
-    PFN_PAL_EXCEPTION_FILTER pfnFilter,
-    void *pvParam,
-    BOOL *pfExecuteHandler);
-
-//
-// Possible results from PAL_TryExcept:
-//
-//   returned exception  pfExecuteHandler  means
-//   ------------------  ----------------  ----------------------------------------
-//   NULL                any               No exception escaped from the try block.
-//   non-NULL            FALSE             An exception escaped from the try block,
-//                                           but the filter did not want to handle it.
-//   non-NULL            TRUE              An exception escaped from the try block,
-//                                           and the filter wanted to handle it.
-//
-
-#define DEBUG_OK_TO_RETURN_BEGIN(arg)
-#define DEBUG_OK_TO_RETURN_END(arg)
-
-#ifdef _PPC_
-// This function does not do anything.  It is just here to be called by
-// PAL_TRY, so we can avoid the body of PAL_TRY being translated by the
-// compiler into a leaf function (i.e., one that does not set up its
-// own frame), because on a hardware fault in a leaf function, we would
-// get a stack that would not be unwindable.
-PALIMPORT VOID PALAPI PAL_DummyCall();
-#define PAL_DUMMY_CALL PAL_DummyCall();
-#else
-#define PAL_DUMMY_CALL
-#endif
-
 #ifdef __cplusplus
-class PAL_CatchHolder
-{
-public:
-    PAL_CatchHolder(_Unwind_Exception *exceptionObject)
-    {
-        __cxa_begin_catch(exceptionObject);
-    }
-
-    ~PAL_CatchHolder()
-    {
-        __cxa_end_catch();
-    }
-};
 
-class PAL_ExceptionHolder
+struct PAL_SEHException
 {
-private:
-    _Unwind_Exception *m_exceptionObject;
 public:
-    PAL_ExceptionHolder(_Unwind_Exception *exceptionObject)
-    {
-        m_exceptionObject = exceptionObject;
-    }
-
-    ~PAL_ExceptionHolder()
-    {
-        if (m_exceptionObject)
-        {
-            _Unwind_DeleteException(m_exceptionObject);
-        }
-    }
-
-    void SuppressRelease()
+    // Note that the following two are actually embedded in this heap-allocated
+    // instance - in contrast to Win32, where the exception record would usually
+    // be allocated on the stack.  This is needed because foreign cleanup handlers
+    // partially unwind the stack on the second pass.
+    EXCEPTION_POINTERS ExceptionPointers;
+    EXCEPTION_RECORD ExceptionRecord;
+    CONTEXT ContextRecord;
+
+    PAL_SEHException(EXCEPTION_RECORD *pExceptionRecord, CONTEXT *pContextRecord)
     {
-        m_exceptionObject = NULL;
+        ExceptionPointers.ExceptionRecord = &ExceptionRecord;
+        ExceptionPointers.ContextRecord = &ContextRecord;
+        ExceptionRecord = *pExceptionRecord;
+        ContextRecord = *pContextRecord;
     }
 };
 
-class PAL_NoHolder
-{
-public:
-    void SuppressRelease() {}
-};
 #endif // __cplusplus
 
 #define PAL_TRY(__ParamType, __paramDef, __paramRef)                            \
```

Comment:
Could you please add a bit more comments describing all of these macros, their parameters and usage? Thank you.